unidist/core/base/utils.py

```diff
@@ -4,27 +4,38 @@
 
 """Utilities used to initialize execution backend."""
 
-import os
-
+from unidist.config import Backend, CpuCount
 from .backend import BackendProxy
 
 
-def init_backend():
+def init_backend(backend=None, num_cpus=None):
     """
     Initialize an execution backend.
 
+    Parameters
+    ----------
+    backend : str, optional
+        Backend that unidist should run on.
+    num_cpus : int, optional
+        Number of CPUs that should be used by backend.
+
     Notes
     -----
-    The concrete execution backend is chosen in depend on
-    `UNIDIST_BACKEND` environment variable.
-    If the variable is not set, Ray backend is used.
+    The concrete execution backend can also be chosen via
+    `UNIDIST_BACKEND` environment variable or ``Backend`` config value.
+    Ray backend is used by default.
     """
-    backend_name = os.environ.get("UNIDIST_BACKEND", "Ray")
+    if num_cpus is not None:
+        CpuCount.put(num_cpus)
```

Comment:
I don't see an issue in checking `num_cpus` in backends. Someone might want to use a backend separately without API layer.

Reply:
I don't see the advantages from if-else branches for `CpuCount`/`Backend`. This is added surplus logic. `None` value processing of parameter can be moved to `Parameter.put` implementation. In that case you can just do something the next:
```python
def init_backend(backend=None, num_cpus=None):
    CpuCount.put(num_cpus)
    Backend.put(backend)

    if Backend.get() == "Ray":
        ...
        initialize_ray(num_cpus=CpuCount.get())
```
Also, in that case you will be able to exclude ` .. =num_cpus or CpuCount.get()` occurrences in the code.

Reply:
I got what you mean but we can't do in this way because a user may set an env var value to None so we intentionally use object() as an unset value [here](https://github.com/modin-project/unidist/pull/21/files#diff-abbd506f028d8566521dd6455c5c3249a7330dffaba2e640fc828de9703c9924R73). Now I am more tend to provide to users env vars, class config and parameters of the CLI interface to specify specific values for unidist behavior tuning.

Reply:
If we suppose, that `None` is one of possible values for some parameter, we should set default value for function parameter as `unidist.parameter._UNSET` :

```python
unidist/api.py
from unidist.parameter import _UNSET as default
def init(backend=default, num_cpus=default ..)
```
In this case, we can use approach, described in the https://github.com/modin-project/unidist/pull/21#discussion_r751118881 above. User can provide `None` values with API/Environment for Parameters, which supports `None` value (`verify` function returns True).

As I see, `None` is a valid value for `str` and `ExactStr` types of parameters (in that cases `verify` returns True for any value). This types is used for `Backend`, `RayRedisAddress`, `RayRedisPassword`. Is `None` expected value for those parameters? If not, it should be processed on `Parameter` level in `put` method - for values from API, in `get` - for values from environment.

Using this implementation I don't see the restrictions to support parameters on API layer of `init` func, which you removed in ba17865907727b0f4c5a74883d6abc816bff1390.

Reply:
`None` is not a valid value for `str` and `ExactStr`. The former is just a case-insensitive string whereas the latter a case-sensitive string.

Since support for parameters of `init` function is a separate task that we have now #30 let's do that there.

Reply:
`ExactStr` can accept `None` value (or any value), because it doesn't have any str-specific methods like `str` type has or other checks. Example RayRedisAddress has type `ExactStr` below:
```python
In [4]: from unidist.config import RayRedisAddress
In [5]: RayRedisAddress.put(None)
In [6]: RayRedisAddress.get() is None
Out[6]: True
```

If `None` isn't supported value for `ExactStr` additional check should be added to exclude this behavior.

I agrre with adding parameters for `init*` functions in a separate #30

Reply:
`ExactStr` implies that a value must be an exact string. If a user passes a different value for the parameter, he doesn't follow the interface but yes it might be good to add a check on the type.

Reply:
added